Add `gpytorch.metrics` and tests

patel-zeel opened the pull request:
As discussed in #1857 with @gpleiss.

Added the following workflows with tests
* `gpytorch.metrics.coverage_error`
* `gpytorch.metrics.mean_standardized_log_loss`
* `gpytorch.metrics.negatave_log_predictive_density`

Initially, I thought of the following version:
`metric(model, likelihood, test_x, test_y)`

But, I had to drop the idea due to the following reason:
* Prediction strategy may differ for each model using `model` and `likelihood`. For example, use of `gpytorch.settings.fast_pred_var()` may not be desired for all models.

Thus, currently, we have the following version suitable for all models:
`metric(pred_dist, test_y)`

Please let me know if corrections/improvements are required.

wjmaddox commented on gpytorch/metrics/metrics.py:
```diff
@@ -0,0 +1,46 @@
+    test_y: torch.Tensor,
+) -> float:
+    """
+    Reference: GPML book
```
Can you put a proper citation in here?

patel-zeel replied:
Sure, I will do that.

patel-zeel replied:
Done.

patel-zeel commented on gpytorch/metrics/metrics.py:
```diff
@@ -0,0 +1,46 @@
+) -> float:
+    """
+    Coverage error for 95% confidence intervals.
+    TODO: Find a good reference to improvise this metric.
```
Thank you for the reference @wjmaddox. 
* What can be considered as a good step size for discretizing [0, 1] space especially for this implementation?
* Shall we have two metrics in this case? 1) `average_coverage_error` 2) `coverage_error`.

wjmaddox replied:
we probably want to have a user option for the number of bins with the default as 10 (or 20), which is what i've used in the past when computing these types of metrics.

re the second point: yes, I think that makes sense, but maybe `coverage_error` should be `percentile_coverage_error`?

patel-zeel replied:
> we probably want to have a user option for the number of bins with the default as 10 (or 20), which is what i've used in the past when computing these types of metrics.

Sounds good! 
> re the second point: yes, I think that makes sense, but maybe `coverage_error` should be `percentile_coverage_error`?

Yeah, explicit is better than implicit. Let's go with `percentile_coverage_error`.

patel-zeel commented on gpytorch/metrics/metrics.py:
```diff
@@ -0,0 +1,46 @@
+        lower, upper = pred_dist.confidence_region()
+        n_samples_within_bounds = ((test_y > lower) * (test_y < upper)).sum()
+        fraction = (n_samples_within_bounds / test_y.shape[0]).item()
+        return abs(0.9545 - fraction)
```
If we proceed with [this](https://github.com/cornellius-gp/gpytorch/pull/1870#discussion_r799398315), the constant will be removed else I'll mention that this constant is the fraction of coverage for `mu` ± 2 * `std`. Does that sound okay?

wjmaddox replied:
thanks for that clarification.

patel-zeel commented:
* Renamed `coverage_error` to `percentile_coverage_error` which takes an argument `percentile` in range [0, 100]. default is 95.
* Added `average_coverage_error` with default `n_bins` to 20.
* Added tests for `average_coverage_error`.

wjmaddox commented on gpytorch/metrics/metrics.py:
```diff
@@ -0,0 +1,73 @@
+        return 0.5 * (torch.log(2 * pi * f_var) + torch.square(test_y - f_mean) / (2 * f_var)).mean().item()
+
+
+def percentile_coverage_error(
```
Change percentiles to quantiles as that's more correct, but this looks good.

patel-zeel replied:
Done.

wjmaddox commented on gpytorch/metrics/metrics.py:
```diff
@@ -0,0 +1,73 @@
+        deviation = standard_normal.icdf(torch.tensor(0.5 + 0.5 * (percentile / 100)))
+        lower = pred_dist.mean - deviation * pred_dist.stddev
+        upper = pred_dist.mean + deviation * pred_dist.stddev
+        n_samples_within_bounds = ((test_y > lower) * (test_y < upper)).sum()
```
sum over the batch dimensions (e.g. sum(-1)) here so that this supports the batched case.

patel-zeel replied:
If I understand this right, in the case where test_y is `n×d`, we want the output of this metric `1×d`? Currently, it outputs a scaler as torch sums over all the dimensions by default.

wjmaddox replied:
Yes, that's exactly the goal here.

wjmaddox replied:
It often makes more sense to consider the metrics for each output separately.

patel-zeel replied:
Yeah, makes perfect sense! Added the batch support.

wjmaddox commented on gpytorch/metrics/metrics.py:
```diff
@@ -0,0 +1,73 @@
+    bins = torch.linspace(0, 100, n_bins + 1)[1:-1]
+    res = 0
+    for percentile in bins:
+        res += percentile_coverage_error(pred_dist, test_y, percentile)
```
change these to quantile

patel-zeel replied:
Done

wjmaddox commented on gpytorch/metrics/metrics.py:
```diff
@@ -0,0 +1,73 @@
+    Average Coverage error.
+
+    """
+    bins = torch.linspace(0, 100, n_bins + 1)[1:-1]
```
```python
torch.linspace(100 / n_bins, 100 - 100 / n_bins, n_bins-1)
```

patel-zeel replied:
Thanks for the suggestion! applied.

wjmaddox commented on test/metrics/test_metrics.py:
```diff
@@ -0,0 +1,80 @@
+
+        model.train()
+        likelihood.train()
+        for i in range(50):
```
Will this pass if it's 2-3 steps (or even 1)?

patel-zeel replied:
All metrics except `quantile_coverage_error` are passing step=1. Currently, I reduced 50 to 20.

wjmaddox replied:
Will this work at 5 or 10 steps?

wjmaddox commented:
This looks pretty much ready other than the nit above. Do you mind adding a notebook describing the usage to `00_Basic_Usage`?

patel-zeel commented:
Sure @wjmaddox! I welcome your suggestions on the notebook.

wjmaddox commented on gpytorch/metrics/metrics.py:
```diff
@@ -0,0 +1,73 @@
+    with torch.no_grad():
+        f_mean = pred_dist.mean
+        f_var = pred_dist.variance
+        return 0.5 * (torch.log(2 * pi * f_var) + torch.square(test_y - f_mean) / (2 * f_var)).mean().item()
```
```suggestion
        return 0.5 * (torch.log(2 * pi * f_var) + torch.square(test_y - f_mean) / (2 * f_var)).mean()
```

patel-zeel replied:
To enable differentiability, shall we remove `torch.no_grad()` as well?

wjmaddox replied:
oops, yes.

wjmaddox commented on gpytorch/metrics/metrics.py:
```diff
@@ -0,0 +1,73 @@
+        upper = pred_dist.mean + deviation * pred_dist.stddev
+        n_samples_within_bounds = ((test_y > lower) * (test_y < upper)).sum(0)
+        fraction = n_samples_within_bounds / test_y.shape[0]
+        return abs(fraction - quantile / 100)
```
```suggestion
        return torch.abs(fraction - quantile / 100)
```

wjmaddox commented on test/metrics/test_metrics.py:
```diff
@@ -0,0 +1,80 @@
+)
+from gpytorch.models import ExactGP
+
+N_PTS = 50
```
```suggestion
N_PTS = 10
```

wjmaddox commented on gpytorch/metrics/metrics.py:
```diff
@@ -0,0 +1,73 @@
+):
+
+    with torch.no_grad():
+        return -pred_dist.log_prob(test_y).item()
```
```suggestion
        return -pred_dist.log_prob(test_y)
```

wjmaddox replied:
Also, would suggest returning a normalization by `test_y.shape[-1]` here as well

patel-zeel replied:
Does that mean normalizing by the number of samples? Also, batch support would be desirable here?

wjmaddox replied:
yes, normalizing by number of samples. this should return a batch if test_y is batched already.

patel-zeel replied:
I am trying to create a separate test case for MultiTaskGP where we can have `test_y` batched. It seems that NLPD returns a scaler in this case. I'm not sure if it is obvious because of my ignorance of Multi-Task GPs. The model I used is the following from [this notebook](https://docs.gpytorch.ai/en/stable/examples/03_Multitask_Exact_GPs/Multitask_GP_Regression.html):

```py
class ExactMultiTaskGPModel(ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super().__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.MultitaskMean(
            gpytorch.means.ConstantMean(), num_tasks=2
        )
        self.covar_module = gpytorch.kernels.MultitaskKernel(
            gpytorch.kernels.RBFKernel(), num_tasks=2, rank=1
        )

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultitaskMultivariateNormal(mean_x, covar_x)
```

wjmaddox replied:
Resolved by updated unit test and the following stack of commits.

wjmaddox reviewed (changes requested):
Sorry for the long delay. This looks good with one suggestion. 

What do you think about adding in mean absolute error and mean square error in here as well?

The suggested are all basically just asking for returning tensors so that the metrics are differentiable which might be useful someday.

patel-zeel commented:
@wjmaddox mean absolute error and mean squared error are great suggestions!

* Do we want to return the metric of batch output size here?
* For the mean squared error, how about having `squared` argument like [Scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_squared_error.html)? `mean_squared_error(pred_dist, test_y, squared=True)`

wjmaddox commented:
> Do we want to return the metric of batch output size here?

Again, we probably want to leave the leading batch dimension and not reduce over it.

> For the mean squared error, how about having squared argument like [Scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_squared_error.html)? mean_squared_error(pred_dist, test_y, squared=True)

Great idea!

patel-zeel commented:
@wjmaddox I have added the metrics and additional tests. I have also added the entries in `Metrics.ipynb`. For some reason, 50% of the time, tests were failing on my local system while keeping N_PTS=10 and the number of epochs=5. To summarize:

* Added `mean_squared_error` and `mean_absolute_error` and their tests.
* Added batch_y testing using MultiTaskGP example.

wjmaddox commented on test/metrics/test_metrics.py:
```diff
@@ -0,0 +1,130 @@
+import math
```
```suggestion
import math
import unittest
from gpytorch.likelihoods import GaussianLikelihood, MultitaskGaussianLikelihood
from gpytorch.likelihoods.multitask_gaussian_likelihood import MultitaskGaussianLikelihood
from gpytorch.means import ConstantMean, MultitaskMean
from gpytorch.kernels import ScaleKernel, RBFKernel, MultitaskKernel
from gpytorch.distributions import MultivariateNormal, MultitaskMultivariateNormal

import torch

import gpytorch
from gpytorch.metrics import (
    # average_coverage_error,
    mean_absolute_error,
    mean_squared_error,
    mean_standardized_log_loss,
    negative_log_predictive_density,
    quantile_coverage_error,
)
from gpytorch.models import ExactGP


class ExactGPModel(ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super().__init__(train_x, train_y, likelihood)
        self.mean_module = ConstantMean()
        self.covar_module = ScaleKernel(RBFKernel())

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return MultivariateNormal(mean_x, covar_x)


class ExactMultiTaskGPModel(ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super().__init__(train_x, train_y, likelihood)
        self.mean_module = MultitaskMean(ConstantMean(), num_tasks=2)
        self.covar_module = MultitaskKernel(RBFKernel(), num_tasks=2, rank=1)

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return MultitaskMultivariateNormal(mean_x, covar_x)


class TestMetricsSingleTask(unittest.TestCase):
    def setUp(self, batch_shape=None):
        train_x, train_y = self.create_data_and_labels(batch_shape=batch_shape)
        untrained_model = self.create_model(train_x, train_y)
        self.trained_model = self.train_model(untrained_model, train_x, train_y)
        self.untrained_model = self.create_model(train_x, train_y)

        self.testing_pts = self.create_data_and_labels(batch_shape=batch_shape,is_training=False)
        self.train_x = train_x
        self.train_y = train_y

    def create_model(self, train_x, train_y):
        likelihood = GaussianLikelihood(batch_shape=train_x.shape[:-2])
        model = ExactGPModel(train_x, train_y, likelihood)
        # enforce determinism
        return model

    def create_data_and_labels(self, n=20, batch_shape=None, seed=0, is_training=True):
        input_range = [0., 1.] if is_training else [1., 1.2]

        torch.random.manual_seed(seed)

        if batch_shape is None:
            batch_shape = torch.Size()
        inputs = torch.linspace(*input_range, n).view(-1, 1).contiguous()
        if batch_shape != torch.Size():
            inputs = inputs.unsqueeze(0)
            inputs = inputs.repeat(*batch_shape, 1, 1)

        labels = torch.sin(inputs * (2 * math.pi)) + torch.randn_like(inputs) * 0.1
        return inputs, labels.squeeze()

    def get_metric(self, model, likelihood, train_data, labels, metric, **kwargs):
        model.eval()
        likelihood.eval()
        pred_dist = likelihood(model(train_data))
        return metric(pred_dist, labels, **kwargs)

    def train_model(self, model, train_x, train_y):
        mll = gpytorch.mlls.ExactMarginalLogLikelihood(model.likelihood, model)
        optimizer = torch.optim.Adam(model.parameters(), lr=0.1)

        model.train()
        model.likelihood.train()
        for _ in range(10):
            optimizer.zero_grad()
            output = model(train_x)
            loss = -mll(output, train_y).sum()
            loss.backward()
            optimizer.step()
        
        return model

    def check_metric(self, metric, **kwargs):
        init_value = self.get_metric(self.untrained_model, self.untrained_model.likelihood, *self.testing_pts, metric, **kwargs)
        final_value = self.get_metric(self.trained_model, self.trained_model.likelihood, *self.testing_pts, metric, **kwargs)
        return init_value, final_value

    def _test_metric(self, metric, should_check_differentiable=True, **kwargs):
        init_value, final_value = self.check_metric(metric, **kwargs)
        self.assertEqual(final_value.shape, self.train_y.shape[:-1])
        self.assertTrue(torch.all(final_value <= init_value))
        if should_check_differentiable:
            self.assertTrue(final_value.requires_grad)  # check that the metric is differentiable

    def test_mean_absolute_error(self):
        self._test_metric(mean_absolute_error)

    def test_mean_squared_error(self):
        self._test_metric(mean_squared_error)

    def test_negative_log_predictive_density(self):
        self._test_metric(negative_log_predictive_density)

    def test_mean_standardized_log_loss(self):
        self._test_metric(mean_standardized_log_loss)

    def test_quantile_coverage_error(self):
        self._test_metric(quantile_coverage_error, should_check_differentiable=False, quantile=95.)

    # def test_average_coverage_error(self):
    #     # note that we compute the coverage error on the training data, which can be susceptible to 
    #     # interpolation
    #     self._test_metric(average_coverage_error, should_check_differentiable=False, n_bins=2)


class TestMetricsBatchedSingleTask(TestMetricsSingleTask):
    def setUp(self):
        batch_shape = torch.Size((4,))
        super().setUp(batch_shape=batch_shape)


class TestMetricsMultiTask(TestMetricsSingleTask):
    def create_data_and_labels(self, n=20, batch_shape=None, seed=0, is_training=True):
        input_range = [0., 1.] if is_training else [1., 1.2]
        torch.random.manual_seed(seed)

        if batch_shape is None:
            batch_shape = torch.Size()
        inputs = torch.linspace(*input_range, n).view(-1, 1).contiguous()
        if batch_shape != torch.Size():
            inputs = inputs.unsqueeze(0)
            inputs = inputs.repeat(*batch_shape, 1, 1)

        labels = torch.cat(
            [
                torch.sin(inputs * (2 * math.pi)) + torch.randn_like(inputs) * 0.1,
                torch.cos(inputs * (2 * math.pi)) + torch.randn_like(inputs) * 0.1,
            ],
            -1
        )
        return inputs, labels.squeeze().contiguous()

    def create_model(self, train_x, train_y):
        likelihood = MultitaskGaussianLikelihood(num_tasks=2)
        model = ExactMultiTaskGPModel(train_x, train_y, likelihood)
        return model

    def _test_metric(self, metric, should_check_differentiable=True, **kwargs):
        init_value, final_value = self.check_metric(metric, **kwargs)
        print(init_value, final_value, init_value.mean(), final_value.mean())
        # here the trailing dimension is the number of tasks
        self.assertEqual(final_value.shape, torch.Size((*self.train_y.shape[:-2], self.train_y.shape[-1],)))
        self.assertTrue(torch.all(final_value.mean() <= init_value.mean()))
        if should_check_differentiable:
            self.assertTrue(final_value.requires_grad)  # check that the metric is differentiable

    def test_negative_log_predictive_density(self):
        init_value, final_value = self.check_metric(negative_log_predictive_density)
        self.assertEqual(final_value.shape, self.train_y.shape[:-2])
        self.assertTrue(torch.all(final_value <= init_value))
        self.assertTrue(final_value.requires_grad)  # check that the metric is differentiable


class TestMetricsBatchedMultiTask(TestMetricsMultiTask):
    def setUp(self):
        batch_shape = torch.Size((4,))
        super().setUp(batch_shape=batch_shape)
```

wjmaddox replied:
This refactors the unit test to properly take into account the following four models: [single task, batched single task, multi-task, and batched multi-task].

wjmaddox commented on gpytorch/metrics/metrics.py:
```diff
@@ -0,0 +1,96 @@
+from math import pi
```
```suggestion
from math import pi

import torch

from ..distributions import MultivariateNormal, MultitaskMultivariateNormal

pi = torch.tensor(pi)


def mean_absolute_error(
    pred_dist: MultivariateNormal,
    test_y: torch.Tensor,
):
    """
    Mean Absolute Error.

    """
    combine_dim = -2 if isinstance(pred_dist, MultitaskMultivariateNormal) else -1
    return torch.abs(pred_dist.mean - test_y).mean(dim=combine_dim)


def mean_squared_error(
    pred_dist: MultivariateNormal,
    test_y: torch.Tensor,
    squared: bool = True,
):
    """
    Mean Squared Error.

    """
    combine_dim = -2 if isinstance(pred_dist, MultitaskMultivariateNormal) else -1
    print(pred_dist.mean.shape, test_y.shape)
    res = torch.square(pred_dist.mean - test_y).mean(dim=combine_dim)
    if not squared:
        return res ** 0.5
    return res


def negative_log_predictive_density(
    pred_dist: MultivariateNormal,
    test_y: torch.Tensor,
):
    combine_dim = -2 if isinstance(pred_dist, MultitaskMultivariateNormal) else -1
    return -pred_dist.log_prob(test_y) / test_y.shape[combine_dim]


def mean_standardized_log_loss(
    pred_dist: MultivariateNormal,
    test_y: torch.Tensor,
):
    """
    Mean Standardized Log Loss.

    Reference: Page No. 23,
    Gaussian Processes for Machine Learning,
    Carl Edward Rasmussen and Christopher K. I. Williams,
    The MIT Press, 2006. ISBN 0-262-18253-X
    """
    combine_dim = -2 if isinstance(pred_dist, MultitaskMultivariateNormal) else -1

    f_mean = pred_dist.mean
    f_var = pred_dist.variance
    return 0.5 * (torch.log(2 * pi * f_var) + torch.square(test_y - f_mean) / (2 * f_var)).mean(dim=combine_dim)


def quantile_coverage_error(
    pred_dist: MultivariateNormal,
    test_y: torch.Tensor,
    quantile: float = 95.,
):
    """
    Quantile coverage error.

    """

    if quantile <= 0 or quantile >= 100:
        raise NotImplementedError("Quantile must be between 0 and 100")

    combine_dim = -2 if isinstance(pred_dist, MultitaskMultivariateNormal) else -1

    standard_normal = torch.distributions.Normal(loc=0.0, scale=1.0)
    deviation = standard_normal.icdf(torch.as_tensor(0.5 + 0.5 * (quantile / 100)))
    lower = pred_dist.mean - deviation * pred_dist.stddev
    upper = pred_dist.mean + deviation * pred_dist.stddev
    n_samples_within_bounds = ((test_y > lower) * (test_y < upper)).sum(combine_dim)
    fraction = n_samples_within_bounds / test_y.shape[combine_dim]
    return torch.abs(fraction - quantile / 100)
```

wjmaddox commented on gpytorch/metrics/metrics.py:
```diff
@@ -0,0 +1,96 @@
+    return torch.abs(fraction - quantile / 100)
+
+
+def average_coverage_error(
```
This is actually summing over all quantiles and not computing an average over each bin, so let's remove this function for now.

wjmaddox reviewed (commented):
I realized I should have taken a deeper look before now. But, these should be all that we need to address (modulo checking for an assert error in the quantile function).

patel-zeel commented:
Thanks a lot for your help in resolving the issues @wjmaddox. 

> modulo checking for an assert error in the quantile function
Is this something that still needs to be addressed? If so, can you please elaborate a bit?

wjmaddox commented:
By that I mean putting a unit test to ensure that the error is raised when we think it does. I'll go ahead and do it.

patel-zeel commented:
Great! Thank you @wjmaddox

wjmaddox commented on test/metrics/test_metrics.py:
```diff
@@ -0,0 +1,189 @@
+    def test_mean_standardized_log_loss(self):
+        self._test_metric(mean_standardized_log_loss)
+
+    def test_quantile_coverage_error(self):
```
```suggestion
    def test_quantile_coverage_error(self):
        self._test_metric(quantile_coverage_error, should_check_differentiable=False, quantile=95.)
        # check that negative or very positive quantile coverages raise errors
        with self.assertRaises(NotImplementedError):
            self.get_metric(
                self.untrained_model,
                self.untrained_model.likelihood,
                *self.testing_pts,
                quantile_coverage_error,
                quantile=-1.,
            )
        with self.assertRaises(NotImplementedError):
            self.get_metric(
                self.untrained_model,
                self.untrained_model.likelihood,
                *self.testing_pts,
                quantile_coverage_error,
                quantile=1000.,
            )
```

wjmaddox commented:
Sorry about the linting disasters previously.

wjmaddox commented on test/metrics/test_metrics.py:
```diff
@@ -0,0 +1,207 @@
+                quantile_coverage_error,
+                quantile=1000.0,
+            )
+        self._test_metric(quantile_coverage_error, should_check_differentiable=False, quantile=95.0)
```
```suggestion
```

wjmaddox commented:
ugh, sorry @patel-zeel can you fix the lint, this is pretty much impossible to do in browser mode. hopefully I didnt break the docs somehow with that commit.

patel-zeel commented:
Done, @wjmaddox. It seems that everything is working.
